src/main/java/com/celements/cleverreach/util/SynthonCssInliner.java

```diff
@@ -0,0 +1,160 @@
+    configs = (configs == null) ? Collections.emptyMap() : configs;
+    LOGGER.trace("Applying the following CSS [{}] to HTML [{}]", css, html);
+    try {
+      String encodedHtml = merge(html, css);
+      if (!configs.containsKey(CONFIG_KEY_SERVER_SECRET)) {
+        throw new CssInlineException(encodedHtml, new RuntimeException("Server secret for synthon "
+            + "serveris missing in config"));
+      }
+      String postData = "secret=" + getHash(configs.get(CONFIG_KEY_SERVER_SECRET) + encodedHtml)
+          + "&html=" + URLEncoder
+              .encode(encodedHtml, "UTF-8");
+      if (configs.containsKey(CONFIG_KEY_REMOVE_CLASSES)) {
+        postData = "removeClasses=" + configs.get(CONFIG_KEY_REMOVE_CLASSES) + "&" + postData;
+      }
+      String result = inlineCss(postData.getBytes(), configs);
+      LOGGER.trace("HTML with CSS INLINED [{}]", result);
+      Inlined inlinedResult = new ObjectMapper().readValue(result, Inlined.class);
+      if (inlinedResult.success) {
+        return inlinedResult.data;
+      }
+      throw new CssInlineException(html, new RuntimeException(inlinedResult.error));
+    } catch (IOException excp) {
+      LOGGER.warn("Failed to apply CSS [{}] to HTML [{}]", css, html, excp);
+      throw new CssInlineException(html, excp);
+    }
+  }
+
+  private String getHash(String inStr) {
+    return Hashing.sha256().hashString(inStr, StandardCharsets.UTF_8).toString();
+  }
+
+  private String inlineCss(byte[] postData, Map<String, String> configs) throws IOException,
+      MalformedURLException,
+      ProtocolException {
+    HttpURLConnection conn = prepareConnection(configs);
+    try (OutputStream out = conn.getOutputStream()) {
+      out.write(postData, 0, postData.length);
+    }
+    StringBuilder resultBuilder = new StringBuilder();
+    try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn
+        .getInputStream()))) {
+      String line = null;
+      while ((line = reader.readLine()) != null) {
+        resultBuilder.append(line);
+      }
+    }
+    String result = resultBuilder.toString();
+    return result;
+  }
+
+  private HttpURLConnection prepareConnection(Map<String, String> configs) throws IOException,
+      MalformedURLException,
+      ProtocolException {
+    if (!configs.containsKey(CONFIG_KEY_INLINE_URL)) {
+      throw new MalformedURLException("URL to inline service is missing in config");
+    }
+    HttpURLConnection conn = (HttpURLConnection) new URL(configs.get(CONFIG_KEY_INLINE_URL))
+        .openConnection();
+    conn.setRequestMethod("POST");
+    conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+    conn.setRequestProperty("Accept", "application/json");
+    conn.setDoOutput(true);
+    return conn;
+  }
+
+  private String merge(String html, String css) {
+    if (!Strings.isNullOrEmpty(css)) {
+      String[] sect;
+      if (html.contains("</head>")) {
+        sect = html.split("(?=</head>)", 2);
```

Comment:
please add a comment to the method that `?=` means "zero-width positive lookahead". I don't think this is a well-known regex feature, but it's a neat one.